pkg/chart/metadata.go

```diff
@@ -68,18 +68,19 @@ type Metadata struct {
 	Type string `json:"type,omitempty"`
 }
 
+// Validate validates the Chart.yaml contents
 func (md *Metadata) Validate() error {
 	if md == nil {
-		return errors.New("metadata is required")
+		return errors.New("chart metadata is required in Chart.yaml")
```

Comment:
It's unclear to me if there's a case where pkg/chart may be validating chart metadata that did NOT come from a Chart.yaml. I think it's safer to remove the "in Chart.yaml" part from the error message, leaving that to downstream consumers. In other words, I'd suggest `pkg/action` should probably report that this error occurred when validating Chart.yaml.

What do you think about that approach?

Reply:
That makes sense. I will make the necessary changes 👍

Reply:
On checking the code, I see that in `pkg/action`, the `Run` gets the chart as a struct. It's already loaded in the `main` package using this line 

https://github.com/helm/helm/blob/895e9192d440e144ead0796fd8c1db81b47ef1c6/cmd/helm/install.go#L177

The `Load` can have errors anywhere. Not sure how we can detect it was in `Chart.yaml`. I was thinking of adding the info about `Chart.yaml` here

https://github.com/helm/helm/blob/ffff0e8c33fde7547ace267c5c91b8c60fc5262c/pkg/chart/loader/load.go#L111

as this function clearly knows that it's loading from files. But then, the `Validate` function is defined on that chart level 

https://github.com/helm/helm/blob/ffff0e8c33fde7547ace267c5c91b8c60fc5262c/pkg/chart/chart.go#L102

and if we add other validations along with metadata validation later in the future, then adding the info that error is in `Chart.yaml` would be wrong. And we can't add the info inside the chart Validate function too, as the consumers need not have used files as the source to construct the chart struct. 

Considering these, I think for now a good option is we can add it here

https://github.com/helm/helm/blob/ffff0e8c33fde7547ace267c5c91b8c60fc5262c/pkg/chart/loader/load.go#L111

But like I said, if we add more validations in `Validate`, then the error message would be misleading. 

What do you think? Did you have any ideas in mind for this? And I need to think about how to append or prepend to an existing error and tell it was from `Chart.yaml` in a sensible way

Reply:
As far as I can see, I don't think you will know what file the metadata has been loaded from.  @bacongobbler maybe you know something more?

Reply:
After looking at different parts of the codebase, this seems like a safe enough assumption to make so my original thinking is a little too defensive. The error change looks good to me.

However, can we make these errors of a typed value such as `ErrChartMetadataMissing`? Have a look at [ErrMissingMetadata](https://github.com/helm/helm/blob/cf79b90867740f58859882b980f3ccc0d84cb44c/pkg/plugin/installer/installer.go#L30-L31) for an example, as well as how that error type is thrown in other areas of the codebase.

By implementing errors in this fashion, error checks like the one on line 121 in pkg/chart/loader/loader_test.go can simply read `if err != ErrChartMetadataMissing`, rather than an exact string match check.